Handle DPI change in wxSplitterWindow

MaartenBent opened the pull request:
Adjust the minimum pane size to the new DPI.
Keep the sash in the same relative position.

MaartenBent committed:
Handle DPI change in wxSplitterWindow

Adjust the minimum pane size to the new DPI.
Keep the sash in the same relative position.

vadz commented on src/generic/splitter.cpp:
```diff
@@ -520,6 +534,14 @@ void wxSplitterWindow::OnSize(wxSizeEvent& event)
+void wxSplitterWindow::OnDPIChanged(wxDPIChangedEvent& event)
+{
+    m_minimumPaneSize = event.ScaleX(m_minimumPaneSize);
+    m_sizeAfterDPIChange = true;
```
Couldn't we just update `m_sashPosition` and `m_lastSize` here too to avoid adding this variable and extra adjustments in `OnSize()`?

MaartenBent replied:
This indeed works fine.

vadz reviewed (commented):
Thanks!

We should merge this if it's the only solution, but I can't help wondering if we could avoid adding the new variable.

MaartenBent committed:
Add a function to check if a DPI event is being processed

MaartenBent committed:
Improve handling DPI change in wxSplitterWindow

wxSplitterWindow can receive multiple size events while the DPI change is being processed.
Keep using the relative sash position while the DPI change is finished.

MaartenBent commented:
This first solution doesn't always work. In the splitter sample it is fine, but in my case the `wxSplitterWindow` gets multiple size events while the DPI change is being processed (maybe because nested splitter windows, or just controls inside controls, I didn't look into it).
After the first call `m_sizeAfterDPIChange` is set to false, so in the next call the normal stuff is used and the sash does not stay at correct position.

The only solution I could think of so far, is to add a `wxNonOwnedWindow::IsDPIChanging` function, which returns true while this code is executed. But this also involves adding an extra variable.

I'll try calculating `m_sashPosition` and `m_lastSize` and in `OnDPIChanged` and see if that will work too.

MaartenBent committed:
Handle DPI change in wxSplitterWindow

MaartenBent commented:
I updated the PR with your suggested solution. The issue I was having was not directly related to `wxSplitterWindow`.

Could you squash everything when merging?

vadz commented:
Great, thanks! I'll squash and merge it soon.

vadz commented:
As mentioned in https://github.com/wxWidgets/wxWidgets/commit/657be8b6dc08bea97bac9cd75c00cb4dd4a2af22#r134617259 this solution doesn't actually work under systems using DIPs, e.g. wxGTK.

It looks wrong to scale values that should be expressed in logical pixels, such as the splitter position, by DPI there.

MaartenBent commented:
On Windows, scaling `m_sashPosition` is definitely needed. I just tested it in the splitter sample.

It looks like `m_sashPosition`/`SetSashPosition` is interpreted as physical pixels.

vadz commented:
> It looks like `m_sashPosition`/`SetSashPosition` is interpreted as physical pixels.

Well, this is the same as LP under MSW anyhow. But under other platforms we need to use LPs and not PPs. So it finally looks like binding `wxSplitterWindow::OnDPIChanged()` only under MSW, i.e. what I did in #24126, might be the correct solution, after all? It is a bit ugly but I don't see how to avoid this.
